### PR TITLE
[metal] uniforms

### PR DESCRIPTION
**Pull request by omni-viral**

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -294,6 +294,11 @@ impl Device {
                 arg.set_data_type(MTLDataType::Texture);
                 arg.set_index(index as NSUInteger);
             }
+            DescriptorType::UniformBuffer | DescriptorType::StorageBuffer => {
+                arg.set_access(MTLArgumentAccess::ReadOnly);
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: I don't think storage buffers are read-only.

Comment (omni-viral) [reply src/backend/metal/src/device.rs]: @msiglreith  Can you explain a bit the difference between `UniformBuffer` and `StorageBuffer`?

Comment (msiglreith) [reply src/backend/metal/src/device.rs]: `UniformBuffer`: Read-only structured buffer
`StorageBuffer`: Read-Write structured buffer (+atomics)

File: src/backend/metal/src/device.rs [hunk 2]
```diff
@@ -835,7 +848,21 @@ impl hal::Device<Backend> for Device {
                                 *old = Some(((new.0).0.clone(), new.1));
                             }
                         },
-                        (&Sampler(_), _) | (&SampledImage(_), _) => panic!("mismatched descriptor set type"),
+                        (&UniformBuffer(ref buffers), Some(&mut n::DescriptorSetBinding::ConstantBuffer(ref mut vec))) |
+                        (&StorageBuffer(ref buffers), Some(&mut n::DescriptorSetBinding::ConstantBuffer(ref mut vec))) => {
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: This should be a separate DescriptorSetBinding (`StorageBuffer`), which needs to be defined on set allocation afaik.

File: src/backend/metal/src/device.rs [hunk 3]
```diff
@@ -752,11 +762,13 @@ impl hal::Device<Backend> for Device {
 
         let mut num_samplers = 0;
         let mut num_textures = 0;
+        let mut num_uniforms = 0;
```
Comment (msiglreith) [inline src/backend/metal/src/device.rs]: I don't know what's the point of these, @kvark

Comment (kvark) [reply src/backend/metal/src/device.rs]: To count resources for allocation in an argument buffer, see code below

File: src/backend/metal/src/device.rs [hunk 4]
```diff
@@ -857,6 +884,26 @@ impl hal::Device<Backend> for Device {
...
+                            mtl_offsets.extend(buffers.iter().map(|buffer| buffer.1.clone()));
+
+                            let encoder: &metal::ArgumentEncoderRef = &encoder;
+
+                            let range = NSRange {
+                                location: offset,
+                                length: mtl_buffers.len() as NSUInteger,
+                            };
+                            unsafe {
+                                msg_send![encoder,
```
Comment (kvark) [inline src/backend/metal/src/device.rs]: can we call it as `encoder.set_buffers` instead of sending a custom message?

Comment (omni-viral) [reply src/backend/metal/src/device.rs]: Should I open PR in metal repo or just add function here?

Comment (kvark) [reply src/backend/metal/src/device.rs]: eh, you are correct. Will leave this for follow up work.

**Pull request merged**